neon_iris/mq_connector.py

```diff
@@ -0,0 +1,141 @@
+# NEON AI (TM) SOFTWARE, Software Development Kit & Application Development System
+# All trademark and other rights reserved by their respective owners
+# Copyright 2008-2021 Neongecko.com Inc.
+# BSD-3
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+# 1. Redistributions of source code must retain the above copyright notice,
+#    this list of conditions and the following disclaimer.
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+# 3. Neither the name of the copyright holder nor the names of its
+#    contributors may be used to endorse or promote products derived from this
+#    software without specific prior written permission.
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+# THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+# CONTRIBUTORS  BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
+# OR PROFITS;  OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE,  EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from time import sleep
+
+import pika.exceptions
+
+from asyncio import Event as AsyncEvent
+from threading import Event, Thread
+from neon_mq_connector.utils.client_utils import MQConnector
+from ovos_utils import LOG
+from pika.adapters.select_connection import SelectConnection
+from pika.channel import Channel
+
+
+class IrisConnector(MQConnector, Thread):
+    async_consumers_enabled = True
+
+    def __init__(self, *args, **kwargs):
+        vhost = kwargs.pop('vhost')
+        # TODO: service_name chosen for backwards-compat. and should be updated
+        #   to something more descriptive
+        kwargs['service_name'] = 'mq_handler'
+        Thread.__init__(self, daemon=True)
+        MQConnector.__init__(self, *args, **kwargs)
+        self.vhost = vhost
+        self._ready = AsyncEvent()
+        self._channel_closed = Event()
+        self._stopping = False
+
+        self._connection = self.init_connection()
+
+    def wait_for_connection(self):
+        LOG.info("Waiting for connection")
+        while not self._ready.is_set():
+            sleep(0.5)
```

Comment:
Sleep isn't thread safe, is it?

Reply:
Ah, probably not. I think the issue I ran into with `run(self._ready.wait())` is that somehow the event would be set, but `wait` would not return. This was really more of a temporary workaround, so I will look into how to fix this

Reply:
Upon further investigation, the issue appears related to the `SelectConnection` using its own IOLoop. I tried passing a specific EventLoop, but that is apparently incompatible with `pika`. I settled on a working solution that still uses `sleep`, but does use the connection's `ioloop` which I believe makes that thread safe..

Looking back, I'm not sure this is functionally different than what was originally there and `sleep` isn't threadsafe, but it probably doesn't have to be in the current implementation since the even't we're checking is threadsafe..